runtime/integration-tests/src/generic/envs/fudge_env/handle.rs

```diff
@@ -0,0 +1,263 @@
+		InherentDataProviders = (
+			FudgeInherentTimestamp,
+			InherentDataProvider,
+			InherentParachain,
+		),
+	>,
+>;
+
+pub type RelayInherentCreator = InherentCreator<
+	RelayBlock,
+	FudgeDummyInherentRelayParachain<RelayHeader>,
+	sp_consensus_babe::inherents::InherentDataProvider,
+>;
+
+pub type ParachainInherentCreator<Block> = InherentCreator<
+	Block,
+	FudgeInherentParaParachain,
+	sp_consensus_aura::inherents::InherentDataProvider,
+>;
+
+pub type DigestCreator<Block> = Box<dyn DigestCreatorT<Block> + Send + Sync>;
+
+pub type RelaychainBuilder<RuntimeApi, Runtime> = fudge::RelaychainBuilder<
+	RelayBlock,
+	RuntimeApi,
+	Runtime,
+	RelayInherentCreator,
+	DigestCreator<RelayBlock>,
+>;
+
+pub type ParachainBuilder<Block, RuntimeApi> = fudge::ParachainBuilder<
+	Block,
+	RuntimeApi,
+	ParachainInherentCreator<Block>,
+	DigestCreator<Block>,
+>;
+
+pub type RelayClient<ConstructApi> = TFullClient<RelayBlock, ConstructApi, TWasmExecutor>;
+pub type ParachainClient<Block, ConstructApi> = TFullClient<Block, ConstructApi, TWasmExecutor>;
+
+pub trait FudgeHandle<T: Runtime> {
+	type RelayRuntime: frame_system::Config<BlockNumber = BlockNumber>
+		+ polkadot_runtime_parachains::paras::Config
+		+ polkadot_runtime_parachains::session_info::Config
+		+ polkadot_runtime_parachains::initializer::Config;
+
+	type RelayConstructApi: ConstructRuntimeApi<
+			RelayBlock,
+			RelayClient<Self::RelayConstructApi>,
+			RuntimeApi = Self::RelayApi,
+		> + Send
+		+ Sync
+		+ 'static;
+
+	type RelayApi: BlockBuilderApi<RelayBlock>
+		+ BabeApi<RelayBlock>
+		+ ParachainHost<RelayBlock>
+		+ ApiExt<RelayBlock, StateBackend = <TFullBackend<RelayBlock> as Backend<RelayBlock>>::State>
+		+ TaggedTransactionQueue<RelayBlock>;
+
+	type ParachainConstructApi: ConstructRuntimeApi<
+			T::Block,
+			ParachainClient<T::Block, Self::ParachainConstructApi>,
+			RuntimeApi = Self::ParachainApi,
+		> + Send
+		+ Sync
+		+ 'static;
+
+	type ParachainApi: BlockBuilderApi<T::Block>
+		+ ApiExt<T::Block, StateBackend = <TFullBackend<T::Block> as Backend<T::Block>>::State>
+		+ AuraApi<T::Block, AuthorityId>
+		+ TaggedTransactionQueue<T::Block>;
+
+	const RELAY_CODE: Option<&'static [u8]>;
+	const PARACHAIN_CODE: Option<&'static [u8]>;
+	const PARA_ID: u32;
+
+	fn relay(&self) -> &RelaychainBuilder<Self::RelayConstructApi, Self::RelayRuntime>;
+	fn relay_mut(&mut self) -> &mut RelaychainBuilder<Self::RelayConstructApi, Self::RelayRuntime>;
+
+	fn parachain(&self) -> &ParachainBuilder<T::Block, Self::ParachainConstructApi>;
+	fn parachain_mut(&mut self) -> &mut ParachainBuilder<T::Block, Self::ParachainConstructApi>;
+
+	fn append_extrinsic(&mut self, chain: Chain, extrinsic: Vec<u8>) -> Result<(), ()>;
+
+	fn with_state<R>(&self, chain: Chain, f: impl FnOnce() -> R) -> R;
+	fn with_mut_state<R>(&mut self, chain: Chain, f: impl FnOnce() -> R) -> R;
+	fn evolve(&mut self);
+
+	fn build(relay_storage: Storage, parachain_storage: Storage) -> Self;
+
+	fn build_relay(
```

Comment:
nit - would rename to `relay_builder` since that's what we're returning.

Reply:
Agree, maybe then the above `build()` method should be also modified to `new_environment()` or something else, WDYT?

Reply:
Yes, or maybe `new` would suffice.